src/data/nat/factorial.lean

```diff
@@ -143,4 +146,43 @@ begin
   { exact add_factorial_succ_le_factorial_add_succ i h }
 end
 
+end factorial
+
+section desc_fac
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation.
+This is closely related to `ring_theory.polynomial.pochhammer`, but much less general. -/
+def desc_fac (n : ℕ) : ℕ → ℕ
+| 0 := 1
+| (k + 1) := (n + k + 1) * desc_fac k
+
+@[simp] lemma desc_fac_zero (n : ℕ) : desc_fac n 0 = 1 := rfl
+
+@[simp] lemma zero_desc_fac (k : ℕ) : desc_fac 0 k = k! :=
+begin
+  induction k with t ht, refl,
+  unfold desc_fac, rw [ht, zero_add, nat.factorial_succ]
+end
+
+lemma desc_fac_succ {n k : ℕ} : desc_fac n k.succ = (n + k + 1) * desc_fac n k := rfl
+
+lemma succ_desc_fac (n : ℕ) : ∀ k, (n + 1) * desc_fac n.succ k = (n + k + 1) * desc_fac n k
+| 0 := by rw [add_zero, desc_fac_zero, desc_fac_zero]
+| (k + 1) := by rw [desc_fac, mul_left_comm, succ_desc_fac, desc_fac, succ_add, ← add_assoc]
+
+/-- `desc_fac n k = (n + k)! / n!`. However, this lemma states a reformulation to avoid ℕ-division.
+See `eval_desc_fac'` if you really need the version that uses ℕ-division. -/
```

Comment:
```suggestion
See `nat.eval_desc_fac'` if you really need the version that uses ℕ-division. -/
```

Reply:
why do we have to put `nat.` here even though we're in `nat`, whilst we can get rid of the `ring_theory.polynomial` above? is the docgen thing that clever?

Reply:
I think you're confusing filenames with namespaces. Doc-gen is not clever enough to infer namespaces, but lean doesn't care about filenames.